### PR TITLE
Add upgrade tests

### PR DESCRIPTION
**Pull request by danielrbradley**
- Set up initial upgrade testing with 2 core resources.
- Use provider attach to make provider debuggable during integration tests.
- Use go:embed instead of printing out source files containing byte arrays.
- Tweak generation to emit ready-compressed files to "./bin" for use within testing.

Fixes https://github.com/pulumi/pulumi-aws-native/issues/114 - so we can run upgrade preview unit tests without credentials in CI.

### CODE REVIEW

File: provider/cmd/pulumi-gen-aws-native/main.go [hunk 1]
```diff
@@ -106,7 +108,12 @@ func main() {
 				panic(fmt.Sprintf("error generating examples: %v", err))
 			}
 			providerDir := filepath.Join(".", "provider", "cmd", "pulumi-resource-aws-native")
-			writePulumiSchema(*supportedSpec, providerDir, "schema.json")
+			if err = writePulumiSchema(*supportedSpec, providerDir, "schema.json"); err != nil {
+				break
```
Comment (mjeffryes) [inline provider/cmd/pulumi-gen-aws-native/main.go:112]: Shouldn't this (and the error cases added here) notify in some when writing fails?

File: provider/go.mod [hunk 2]
```diff
@@ -1,6 +1,8 @@
 module github.com/pulumi/pulumi-aws-native/provider
 
-go 1.21
+go 1.21.0
```
Comment (mjeffryes) [inline provider/go.mod:3]: curious, why pin to 1.21.0?

File: provider/go.mod [hunk 3]
```diff
@@ -183,28 +188,33 @@ require (
...
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 	lukechampine.com/frand v1.4.2 // indirect
 	sourcegraph.com/sourcegraph/appdash v0.0.0-20211028080628-e2786a622600 // indirect
 )
 
 replace github.com/lestrrat-go/jsschema => github.com/mikhailshilkov/jsschema v0.0.0-20210924145243-fc93fd28ee1b
+
+replace github.com/pulumi/pulumi/pkg/v3 => github.com/pulumi/pulumi/pkg/v3 v3.91.1
+
+replace github.com/pulumi/pulumi/sdk/v3 => github.com/pulumi/pulumi/sdk/v3 v3.91.1
```
Comment (mjeffryes) [inline provider/go.mod:220]: I'm confused about this too

File: provider/pkg/provider/provider_2e2_test.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	})
+}
+
+func testUpgradeFrom(test *pulumitest.PulumiTest, version string) {
+	test.T().Helper()
+	result := providertest.PreviewProviderUpgrade(test, "aws-native", version)
+	assertpreview.HasNoChanges(test.T(), result)
+}
+
+func testLifecycle(test *pulumitest.PulumiTest) auto.OutputMap {
```
Comment (mjeffryes) [inline provider/pkg/provider/provider_2e2_test.go:40]: this method is unused for now?

File: provider/pkg/provider/provider_2e2_test.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/pulumi/providertest/pulumitest"
+	"github.com/pulumi/providertest/pulumitest/assertpreview"
+	"github.com/pulumi/providertest/pulumitest/assertrefresh"
+	"github.com/pulumi/providertest/pulumitest/opttest"
+	"github.com/pulumi/pulumi-aws-native/provider/pkg/provider"
+	"github.com/pulumi/pulumi/sdk/v3/go/auto"
+	pulumirpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+)
+
+func TestE2eSnapshots(t *testing.T) {
```
Comment (mjeffryes) [inline provider/pkg/provider/provider_2e2_test.go:20]: love how straightforward this is. :)

File: provider/pkg/provider/provider_2e2_test.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	t.Run("webAcl-0.94.0", func(t *testing.T) {
+		t.Parallel()
+		test := newAwsTest(t, filepath.Join("testdata", "webAcl"))
+		testUpgradeFrom(test, "0.94.0")
+	})
+}
+
+func testUpgradeFrom(test *pulumitest.PulumiTest, version string) {
+	test.T().Helper()
+	result := providertest.PreviewProviderUpgrade(test, "aws-native", version)
```
Comment (mjeffryes) [inline provider/pkg/provider/provider_2e2_test.go:36]: does this helper include all the logic to record state too?

File: provider/pkg/provider/provider_2e2_test.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+		}
+		metadataBytes, err := os.ReadFile(filepath.Join(cmdDir, "metadata.json.gz"))
+		if err != nil {
+			return nil, err
+		}
+		return provider.NewAwsNativeProvider(nil, "aws-native", "0.1.0", schemaBytes, metadataBytes)
+	})
+}
+
+func skipIfShort(t *testing.T) {
```
Comment (mjeffryes) [inline provider/pkg/provider/provider_2e2_test.go:77]: noting that this is also unused for now

**Review by mjeffryes — APPROVED** [review]
A few notes, but nothing blocking once we get the credentials working so tests can pass

File: provider/go.mod (hunk 2 above)
Comment (danielrbradley) [reply provider/go.mod:3]: I didn't make this change - I think this was done by `go mod tidy` 🤷

File: provider/pkg/provider/provider_2e2_test.go (hunk 6 above)
Comment (danielrbradley) [reply provider/pkg/provider/provider_2e2_test.go:36]: Yup - there's options to control the state storage but the defaults will store it in the same place as the old version of upgrade test.

File: provider/cmd/pulumi-gen-aws-native/main.go (hunk 1 above)
Comment (danielrbradley) [reply provider/cmd/pulumi-gen-aws-native/main.go:112]: It does, but the existing code pattern is just a bit weird. When it breaks the loop there's then a check for the error. I don't like the pattern, but just keeping it matching the code around it.

File: provider/go.mod (hunk 3 above)
Comment (danielrbradley) [reply provider/go.mod:220]: When pulling in github.com/pulumi/providertest it seems to force pu/pkg to upgrade, however there's a codegen bug that leave the Go SDK uncompilable. Therefore, we have to use a replacement until the regression is fixed to forcably downgrade pulumi/pkg.

File: provider/pkg/provider/provider_2e2_test.go (hunk 6 above)
Comment (danielrbradley) [reply provider/pkg/provider/provider_2e2_test.go:36]: Oh, here's the docstring for reference:
> PreviewProviderUpgrade captures the state of a stack from a baseline provider configuration, then previews the stack with the current provider configuration. Uses a default cache directory of "testdata/recorded/TestProviderUpgrade/{programName}/{baselineVersion}".

File: provider/pkg/provider/provider_2e2_test.go (hunk 4 above)
Comment (danielrbradley) [reply provider/pkg/provider/provider_2e2_test.go:40]: Yes, until we re-instate the new style integration tests. Will tidy

File: provider/pkg/provider/provider_2e2_test.go (hunk 7 above)
Comment (danielrbradley) [reply provider/pkg/provider/provider_2e2_test.go:77]: Yup, will clean up for now

**Comment by danielrbradley** [comment]
@mjeffryes I think that's all comments addressed. If you can just confirm you're happy with [implement skipping credentials validation](https://github.com/pulumi/pulumi-aws-native/pull/1292/commits/bc785b04f00ec7fdb4071fafa55ec51ee6cf139c) then we're good to go!

File: provider/go.mod (hunk 3 above)
Comment (danielrbradley) [reply provider/go.mod]: New version just out so have upgraded and removed this again 🎉

**Pull request merged**